src/kakarot/accounts/library.cairo

```diff
@@ -160,57 +153,112 @@ namespace AccountContract {
-    // @dev Recursively validates if tx is signed and valid for each call -> see utils/eth_transaction.cairo
+    // @notice Validate the signature of the call in the call array.
+    // @dev This function validates an Ethereum transaction by checking if the transaction
+    // is correctly signed by the given address, and if the nonce in the transaction
+    // matches the nonce of the account. It decodes the transaction using the decode function,
+    // and then verifies the Ethereum signature on the transaction hash.
     // @param call_array_len The length of the call array.
     // @param call_array The call array.
     // @param calldata_len The length of the calldata.
     // @param calldata The calldata.
+    // @param signature_len The length of tx signature
+    // @param signature The tx signature.
+    // @param nonce The nonce of the tx
+    // @param chain_id The chain_id
     func validate{
         syscall_ptr: felt*,
         pedersen_ptr: HashBuiltin*,
         bitwise_ptr: BitwiseBuiltin*,
         range_check_ptr,
-    }(call_array_len: felt, call_array: CallArray*, calldata_len: felt, calldata: felt*) -> () {
+    }(
+        tx: model.EthTransaction*,
+        tx_data_len: felt,
+        tx_data: felt*,
+        signature_len: felt,
+        signature: felt*,
+        outside_nonce: felt,
+        chain_id: felt,
+    ) -> () {
         alloc_locals;
-        if (call_array_len == 0) {
-            // Validates that this account doesn't have code. Check only once at the end of the recursion.
-            let (bytecode_len) = Account_bytecode_len.read();
-            with_attr error_message("EOAs cannot have code") {
-                assert bytecode_len = 0;
-            }
-
-            return ();
-        }
 
         let (address) = Account_evm_address.read();
-        let (tx_info) = get_tx_info();
 
         // Assert signature field is of length 5: r_low, r_high, s_low, s_high, v
-        assert tx_info.signature_len = 5;
-        let r = Uint256(tx_info.signature[0], tx_info.signature[1]);
-        let s = Uint256(tx_info.signature[2], tx_info.signature[3]);
-        let v = tx_info.signature[4];
-        let (_, chain_id) = unsigned_div_rem(tx_info.chain_id, 2 ** 32);
-
-        // Unpack the tx data
-        let packed_tx_data_len = [call_array].data_len;
-        let packed_tx_data = calldata + [call_array].data_offset;
-
-        let tx_data_len = [packed_tx_data];
-        let (tx_data) = Helpers.load_packed_bytes(
-            packed_tx_data_len - 1, packed_tx_data + 1, tx_data_len
-        );
+        with_attr error_message("Incorrect signature length") {
+            assert signature_len = 5;
+        }
+        let r = Uint256(signature[0], signature[1]);
+        let s = Uint256(signature[2], signature[3]);
+        let v = signature[4];
+
+        with_attr error_message("Invalid nonce") {
+            assert tx.signer_nonce = outside_nonce;
```

Comment:
we just need to check that the `tx.signer_nonce == Account_nonce`

Reply:
Done with [f5577ec](https://github.com/kkrt-labs/kakarot/pull/1246/commits/f5577ec16caa482c5b2bd5885ad2c92a736bd727)